compiler/rustc_hir_typeck/src/expr.rs

```diff
@@ -3463,24 +3481,21 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             // up to a depth of three
             None
         } else {
-            field_path.push(candidate_field.ident(self.tcx).normalize_to_macros_2_0());
-            let field_ty = candidate_field.ty(self.tcx, subst);
-            if matches(candidate_field, field_ty) {
+            field_path.push(candidate_field.0);
+            let field_ty = candidate_field.1;
+            if matches(candidate_field.0, field_ty) {
```

Comment:
while you're here, can you slightly cleanup the function?

- change the `debug!` to an `instrument` on the function
- change the `if field_path.len() > 3 {` and `if matches(candidate_field.0, field_ty) {` to early returns to avoid 2 levels of identation